common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +177,463 @@
+                                    .getFullStackTrace(e));
+                }
             }
+        }
 
-            if (extensions.isEmpty()) {
-                throw new EnhancedServiceNotFoundException(
-                    "not found service provider for : " + service.getName() + "[" + activateName
-                        + "] and classloader : " + ObjectUtils.toString(loader));
+        @SuppressWarnings("rawtypes")
+        private S loadExtension(String activateName, ClassLoader loader, Class[] argTypes,
+                                Object[] args) {
+            if (io.seata.common.util.StringUtils.isEmpty(activateName)) {
+                throw new IllegalArgumentException("the name of service provider for [" + type.getName() + "] name is null");
+            }
+            try {
+                loadAllExtensionClass(loader);
+                ExtensionURL cachedExtensionURL = getCachedServiceURL(activateName);
+                return getExtension(cachedExtensionURL, loader, argTypes, args);
+            } catch (Throwable e) {
+                if (e instanceof EnhancedServiceNotFoundException) {
+                    throw (EnhancedServiceNotFoundException)e;
+                } else {
+                    throw new EnhancedServiceNotFoundException(
+                            "not found service provider for : " + type.getName() + " caused by " + ExceptionUtils
+                                    .getFullStackTrace(e));
+                }
             }
-            Class<?> extension = extensions.get(extensions.size() - 1);
-            S result = initInstance(service, extension, argTypes, args);
-            if (!foundFromCache && LOGGER.isInfoEnabled()) {
-                LOGGER.info("load " + service.getSimpleName() + "[" + activateName + "] extension by class[" + extension
-                    .getName() + "]");
+        }
+
+        private S getExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes,
+                               Object[] args) {
+            if (url == null) {
+                throw new EnhancedServiceNotFoundException("service provider not found");
             }
-            return result;
-        } catch (Throwable e) {
-            if (e instanceof EnhancedServiceNotFoundException) {
-                throw (EnhancedServiceNotFoundException)e;
+            if (Scope.SINGLETON.equals(url.getScope())) {
+                Holder<Object> holder = extensionInstances.get(url);
+                if (holder == null) {
+                    extensionInstances.putIfAbsent(url, new Holder<>());
+                    holder = extensionInstances.get(url);
+                }
+                Object instance = holder.get();
+                if (instance == null) {
+                    synchronized (holder) {
+                        instance = holder.get();
+                        if (instance == null) {
+                            instance = createExtension(url, loader, argTypes, args);
```

Comment:
Why not use createNewExtension here directly?

Reply:
Our purpose is to distinguish whether the extension point is Singleton or multiple. When the LoadLevel's prototype was set to singleton, we return the same instance (using cache), while it was set to Prototype, we use CreateNewExtension() to create new instance

Reply:
In the 415 row, use createNewExtension instead of createExtension.